### ISSUE TITLE
Transactional Producer instance gets timeout getting commit lock while second instance starts 

### ISSUE THREAD

**Issue by PatrickChauveau** [confluentinc/parallel-consumer#803]
Hi,

While instance A is consuming/producing records, instance B is started, the following sequence takes place:

1. time 0: 	
- Instance B requests joining group, at this time several timeout clock start.
- Instance A stops consuming and produces message _Request joining group due to: group is already rebalancing_ every 3sec.
2. time 0 + max.poll.interval.ms: 
- Instance B produces message _JoinGroup failed: The coordinator is not aware of this member_
- Instance A produces message: _consumer poll timeout has expired. This means the time between subsequent calls to poll() was longer than the configured max.poll.interval.ms_
- Instance B continues requesting joining group and then produces this sequence of messages: 
> Successfully joined group with generation Generation{...}
> Successfully synced group in generation Generation{...}
> Notifying assignor about the new Assignment(partitions=[...])
> Assigned 4 total (4 new) partition(s) [...]
> Setting offset for partition [...]
- Instance B starts consuming
3. time 0 + request.timeout.ms: 
- Instance A outputs message _Resetting the last seen epoch of partition_
4. time 0 + commitLockAcquisitionTimeout: 
- Instance A finally produces stack trace with following messages and is definitely stuck

```
User provided listener io.confluent.parallelconsumer.ParallelEoSStreamProcessor failed on invocation of onPartitionsRevoked for partitions
Caused by: java.util.concurrent.TimeoutException: Timeout getting commit lock (which was set to PT5M). Slow processing or too many records being ack'd? Try increasing the commit lock timeout (commitLockAcquisitionTimeout), or reduce your record processing time.
Error from poll control thread, will attempt controlled shutdown, then rethrow. Error: There is a newer producer with the same transactionalId which fences the current one
```

The used configuration is:

- PC version : 0.5.2.8
- number of partitions : 4
- max.poll.records: 150
- max.poll.interval.ms: 180000
- request.timeout.ms: 210000
- linger.ms = 65
- delivery.timeout.ms: 240000
- allowEagerProcessingDuringTransactionCommit=false, 
- commitLockAcquisitionTimeout=PT5M, 
- produceLockAcquisitionTimeout=PT1M, 
- commitInterval=PT0.1S, 
- ordering=UNORDERED, 
- commitMode=PERIODIC_TRANSACTIONAL_PRODUCER, 
- maxConcurrency=64, 
- invalidOffsetMetadataPolicy=FAIL, 
- defaultMessageRetryDelay=PT1S, 
- retryDelayProvider=null, 
- sendTimeout=PT10S, 
- offsetCommitTimeout=PT10S, 
- batchSize=1, 
- thresholdForTimeSpendInQueueWarning=PT10S, 
- maxFailureHistory=10, 
- shutdownTimeout=PT10S, 
- drainTimeout=PT30S, 
- messageBufferSize=50, 
- initialLoadFactor=2, 
- maximumLoadFactor=100

**Comment by rkolesnev** [comment]
Hi @PatrickChauveau - thanks for report.

Is the issue consistently reproducible? 
Just starting 2nd instance when 1 instance is active is enough to observe it?

Would it be possible for you to create a reproducible example ? Either as a simple app or integration test - as that would help a lot to identify potential bug.

**Comment by rkolesnev** [comment]
Yep - i have found a race condition - it is between commitOffsets on normal work and commitOffsets due to partition revocation. 
We already had added additional checks / guards there - but still have synchronisation issue somewhere leading to deadlock.
Will investigate further to flush it out.

**Comment by PatrickChauveau** [comment]
Glad you found a clue! It does not appear every time, might be 50% occurrences. We set in place health endpoint associated with kubernetes liveness probe to thwart the issue (it needs sometimes 2 or 3 restarts to stabilize). _ParallelEosStreamProcessor.getFailureCause()_ helped a lot for this! Failure message: _Error from poll control thread: There is a newer producer with the same transactionalId which fences the current one._

**Comment by astubbs** [comment]
Tracked in the maintained fork as astubbs/parallel-consumer#44, which carries a summary of this
issue and its current status there.

The **Confluent-hosted version** of this project is no longer maintained - its README says so.
**Parallel Consumer itself continues** in my fork: https://github.com/astubbs/parallel-consumer

Nothing in this repository is closed or changed by this comment.

**Comment by astubbs** [comment]
@PatrickChauveau - a status on this from the maintained fork, where it is tracked as astubbs/parallel-consumer#44. The blocking site is the wait at the top of `onPartitionsRevoked` on the producer's transaction lock, which runs on the poll thread and so is bounded by `max.poll.interval.ms`; overrun it and you get exactly the sequence you reported. [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) bounds that wait by `commitLockAcquisitionTimeout` instead of leaving it unbounded (astubbs/parallel-consumer#466), and its release notes name this as one of two known exceptions to the release's claim, both transactional-mode only. What is not yet fixed: that bound defaults to five minutes, which can still exceed `max.poll.interval.ms`, so a long enough commit during a rebalance still evicts the member. Bounding the wait to something under the poll interval is astubbs/parallel-consumer#408, open; the earlier decline-the-lock design was refuted by astubbs/parallel-consumer#466 and is no longer what that PR does. The `commitOffsets` race that @rkolesnev found in 2024 is part of the same mechanism.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
